crates/sui-rest-api/src/transactions/execution.rs

```diff
@@ -342,3 +338,106 @@ fn derive_balance_changes(
         })
         .collect()
 }
+
+pub struct SimulateTransaction;
+
+impl ApiEndpoint<RestService> for SimulateTransaction {
+    fn method(&self) -> axum::http::Method {
+        axum::http::Method::POST
+    }
+
+    fn path(&self) -> &'static str {
+        "/transactions/simulate"
+    }
+
+    fn operation(
+        &self,
+        generator: &mut schemars::gen::SchemaGenerator,
+    ) -> openapiv3::v3_1::Operation {
+        OperationBuilder::new()
+            .tag("Transactions")
+            .operation_id("SimulateTransaction")
+            .request_body(RequestBodyBuilder::new().bcs_content().build())
+            .response(
+                200,
+                ResponseBuilder::new()
+                    .json_content::<TransactionSimulationResponse>(generator)
+                    .bcs_content()
+                    .build(),
+            )
+            .build()
+    }
+
+    fn handler(&self) -> RouteHandler<RestService> {
+        RouteHandler::new(self.method(), simulate_transaction)
+    }
+}
+
+async fn simulate_transaction(
+    State(state): State<Option<Arc<dyn TransactionExecutor>>>,
+    accept: AcceptFormat,
+    //TODO allow accepting JSON as well as BCS
+    Bcs(transaction): Bcs<Transaction>,
+) -> Result<ResponseContent<TransactionSimulationResponse>> {
+    let executor = state.ok_or_else(|| anyhow::anyhow!("No Transaction Executor"))?;
+
+    if transaction.gas_payment.objects.is_empty() {
```

Comment:
Is it necessarily an issue to simulate a transaction without a gas payment? I thought I saw the authority codebase handled that by generating a mock gas object.

Reply:
Yes, so the main reason to support a simulate_transaction api thats different from the resolve one is that the input format is different. Resolve takes a json payload that may have holes in it while this one takes a fully formed Transaction and explicitly doesn't try to fill in any holes. I'll add documentation to that affect.